contracts/FeralfileArtworkV3.sol

```diff
@@ -0,0 +1,433 @@
+        public
+        view
+        returns (uint256)
+    {
+        require(index < totalEditionOfArtwork(artworkID));
+        return allArtworkEditions[artworkID][index];
+    }
+
+    /// @notice A distinct Uniform Resource Identifier (URI) for a given asset.
+    function tokenURI(uint256 tokenId)
+        public
+        view
+        virtual
+        override
+        returns (string memory)
+    {
+        require(
+            _exists(tokenId),
+            "ERC721Metadata: URI query for nonexistent token"
+        );
+
+        string memory baseURI = _tokenBaseURI;
+        if (bytes(baseURI).length == 0) {
+            baseURI = "ipfs://";
+        }
+
+        return
+            string(
+                abi.encodePacked(
+                    baseURI,
+                    artworkEditions[tokenId].ipfsCID,
+                    "/metadata.json"
+                )
+            );
+    }
+
+    /// @notice Update the base URI for all tokens
+    function setTokenBaseURI(string memory baseURI_) external onlyAuthorized {
+        _tokenBaseURI = baseURI_;
+    }
+
+    /// @notice A URL for the opensea storefront-level metadata
+    function contractURI() public view returns (string memory) {
+        return _contractURI;
+    }
+
+    /// @notice Called with the sale price to determine how much royalty
+    //          is owed and to whom.
+    /// @param tokenId - the NFT asset queried for royalty information
+    /// @param salePrice - the sale price of the NFT asset specified by tokenId
+    /// @return receiver - address of who should be sent the royalty payment
+    /// @return royaltyAmount - the royalty payment amount for salePrice
+    function royaltyInfo(uint256 tokenId, uint256 salePrice)
+        external
+        view
+        override
+        returns (address receiver, uint256 royaltyAmount)
+    {
+        require(
+            _exists(tokenId),
+            "ERC2981: query royalty info for nonexistent token"
+        );
+
+        receiver = royaltyPayoutAddress;
+
+        royaltyAmount =
+            (salePrice * secondarySaleRoyaltyBPS) /
+            MAX_ROYALITY_BPS;
+    }
+
+    /// @notice isValidRequest validates a message by ecrecover to ensure
+    //          it is signed by either the contract owner or a trustee
+    //          who is set by the contract owner.
+    /// @param message_ - the raw message for signing
+    /// @param r_ - part of signature for validating parameters integrity
+    /// @param s_ - part of signature for validating parameters integrity
+    /// @param v_ - part of signature for validating parameters integrity
+    function isValidRequest(
+        bytes32 message_,
+        bytes32 r_,
+        bytes32 s_,
+        uint8 v_,
+        uint256 tokenID_
+    ) internal view returns (bool authorized) {
+        address signer = ECDSA.recover(
+            ECDSA.toEthSignedMessageHash(message_),
+            v_,
+            r_,
+            s_
+        );
+        authorized =
```

Comment:
It supposes to validate the signer is `transferParam_.from`.

Reply:
Fixed, changed verify signer is transferParam_.from